File: src/Stage_3_CSR/compare_r_vs_python_results.py
#!/usr/bin/env python3
"""
Compare R and Python CSR/ecosystem services results
"""

import pandas as pd
import numpy as np

print("=" * 80)
print("COMPARISON: R vs Python Implementation Results")
print("=" * 80)
print()

# Load both results
print("Loading data...")
r_results = pd.read_parquet('model_data/outputs/perm2_production/perm2_11680_with_ecoservices_R_20251030.parquet')
py_results = pd.read_parquet('model_data/outputs/perm2_production/perm2_11680_with_ecoservices_20251030.parquet')

print(f"R results: {len(r_results)} species, {r_results.shape[1]} columns")
print(f"Python results: {len(py_results)} species, {py_results.shape[1]} columns")
print()

# Check same species
assert set(r_results['wfo_taxon_id']) == set(py_results['wfo_taxon_id']), "Different species sets!"

# Sort both by taxon_id for comparison
r_results = r_results.sort_values('wfo_taxon_id').reset_index(drop=True)
py_results = py_results.sort_values('wfo_taxon_id').reset_index(drop=True)

print("=" * 80)
print("1. CSR SCORE COMPARISON")
print("=" * 80)
print()

# Compare CSR scores
csr_cols = ['C', 'S', 'R']
for col in csr_cols:
    r_val = r_results[col].values
    py_val = py_results[col].values

    # Both NaN
    both_nan = np.isnan(r_val) & np.isnan(py_val)
    print(f"{col} - Both NaN: {both_nan.sum()} species")

    # Both valid
    both_valid = ~np.isnan(r_val) & ~np.isnan(py_val)
    print(f"{col} - Both valid: {both_valid.sum()} species")

    if both_valid.sum() > 0:
        diff = np.abs(r_val[both_valid] - py_val[both_valid])
        max_diff = diff.max()
        mean_diff = diff.mean()
        print(f"  Max difference: {max_diff:.10f}")
        print(f"  Mean difference: {mean_diff:.10f}")

        if max_diff < 1e-6:
            print(f"  ✓ IDENTICAL (within 1e-6 tolerance)")
        elif max_diff < 0.01:
            print(f"  ✓ VERY CLOSE (within 0.01 tolerance)")
        else:
            print(f"  ⚠ DIFFERENCES DETECTED")

            # Show examples with largest differences
            top_diff_idx = diff.argsort()[-5:]
            print(f"\n  Top 5 differences:")
            for idx in top_diff_idx:
                actual_idx = np.where(both_valid)[0][idx]
                sp_name = r_results.iloc[actual_idx]['wfo_scientific_name']
                r_score = r_val[actual_idx]
                py_score = py_val[actual_idx]
                diff_val = diff[idx]
                print(f"    {sp_name}: R={r_score:.6f}, Py={py_score:.6f}, diff={diff_val:.6f}")
    print()

print("=" * 80)
print("2. NaN SPECIES COMPARISON")
print("=" * 80)
print()

r_nan = r_results[np.isnan(r_results['C'])]
py_nan = py_results[np.isnan(py_results['C'])]

print(f"R NaN species: {len(r_nan)}")
print(f"Python NaN species: {len(py_nan)}")

r_nan_set = set(r_nan['wfo_taxon_id'])
py_nan_set = set(py_nan['wfo_taxon_id'])

if r_nan_set == py_nan_set:
    print("✓ IDENTICAL: Same species fail in both implementations")
    print(f"\nExamples (first 10):")
    for name in list(r_nan['wfo_scientific_name'])[:10]:
        print(f"  - {name}")
else:
    print("⚠ DIFFERENT: Different species fail")
    only_r = r_nan_set - py_nan_set
    only_py = py_nan_set - r_nan_set
    if only_r:
        print(f"\n  Only R fails ({len(only_r)} species):")
        for taxon_id in list(only_r)[:5]:
            print(f"    {taxon_id}")
    if only_py:
        print(f"\n  Only Python fails ({len(only_py)} species):")
        for taxon_id in list(only_py)[:5]:
            print(f"    {taxon_id}")

print()

print("=" * 80)
print("3. ECOSYSTEM SERVICES COMPARISON")
print("=" * 80)
print()

service_cols = [
    'npp_rating', 'decomposition_rating', 'nutrient_cycling_rating',
    'nutrient_retention_rating', 'nutrient_loss_rating',
    'carbon_biomass_rating', 'carbon_recalcitrant_rating',
    'carbon_total_rating', 'erosion_protection_rating',
    'nitrogen_fixation_rating'
]

all_match = True
for col in service_cols:
    if col not in r_results.columns or col not in py_results.columns:
        print(f"⚠ {col}: Missing in one implementation")
        all_match = False
        continue

    match = (r_results[col] == py_results[col]) | (r_results[col].isna() & py_results[col].isna())
    match_pct = 100 * match.sum() / len(r_results)

    if match_pct == 100:
        print(f"✓ {col}: 100% match")
    else:
        print(f"⚠ {col}: {match_pct:.2f}% match ({(~match).sum()} differences)")
        all_match = False

        # Show examples
        diff_idx = np.where(~match)[0][:5]
        print(f"  Examples:")
        for idx in diff_idx:
            sp_name = r_results.iloc[idx]['wfo_scientific_name']
            r_val = r_results.iloc[idx][col]
            py_val = py_results.iloc[idx][col]
            print(f"    {sp_name}: R={r_val}, Py={py_val}")

print()

print("=" * 80)
print("FINAL VERDICT")
print("=" * 80)
print()

# Overall check
csr_match = True
for col in ['C', 'S', 'R']:
    r_val = r_results[col].values
    py_val = py_results[col].values
    both_valid = ~np.isnan(r_val) & ~np.isnan(py_val)
    if both_valid.sum() > 0:
        diff = np.abs(r_val[both_valid] - py_val[both_valid])
        if diff.max() > 0.01:
            csr_match = False

nan_match = set(r_nan['wfo_taxon_id']) == set(py_nan['wfo_taxon_id'])

if csr_match and nan_match and all_match:
    print("✓✓✓ PERFECT MATCH")
    print()
    print("  • CSR scores identical (within floating-point precision)")
    print("  • Same 30 species fail with NaN")
    print("  • All ecosystem services match")
    print()
    print("  R implementation is equivalent to Python implementation")
    print("  Ready for Prof Shipley's review")
elif csr_match and nan_match:
    print("✓✓ CSR MATCH, ECOSYSTEM SERVICES DIFFER")
    print()
    print("  • CSR scores identical")
    print("  • Same edge cases")
    print("  • Review ecosystem service logic for differences")
else:
    print("⚠ DISCREPANCIES FOUND")
    print()
    print("  Review implementation differences above")

print()

File: src/Stage_3_CSR/compare_r_vs_python_results.R
#!/usr/bin/env Rscript
################################################################################
# Compare R and Python CSR/Ecosystem Services Results
################################################################################

suppressPackageStartupMessages({
  library(arrow)
  library(dplyr)
})

cat("============================================================\n")
cat("COMPARISON: R vs Python Implementation Results\n")
cat("============================================================\n\n")

# Load data
cat("Loading data...\n")
r_results <- read_parquet("model_data/outputs/perm2_production/perm2_11680_with_ecoservices_R_20251030.parquet")
py_results <- read_parquet("model_data/outputs/perm2_production/perm2_11680_with_ecoservices_20251030.parquet")

cat(sprintf("R results: %d species, %d columns\n", nrow(r_results), ncol(r_results)))
cat(sprintf("Python results: %d species, %d columns\n", nrow(py_results), ncol(py_results)))
cat("\n")

# Sort both by taxon_id
r_results <- r_results %>% arrange(wfo_taxon_id)
py_results <- py_results %>% arrange(wfo_taxon_id)

# Check same species
stopifnot(identical(r_results$wfo_taxon_id, py_results$wfo_taxon_id))

cat("============================================================\n")
cat("1. CSR SCORE COMPARISON\n")
cat("============================================================\n\n")

for (col in c("C", "S", "R")) {
  r_val <- r_results[[col]]
  py_val <- py_results[[col]]

  both_nan <- is.na(r_val) & is.na(py_val)
  both_valid <- !is.na(r_val) & !is.na(py_val)

  cat(sprintf("%s - Both NaN: %d species\n", col, sum(both_nan)))
  cat(sprintf("%s - Both valid: %d species\n", col, sum(both_valid)))

  if (sum(both_valid) > 0) {
    diff <- abs(r_val[both_valid] - py_val[both_valid])
    max_diff <- max(diff)
    mean_diff <- mean(diff)

    cat(sprintf("  Max difference: %.10f\n", max_diff))
    cat(sprintf("  Mean difference: %.10f\n", mean_diff))

    if (max_diff < 1e-6) {
      cat("  ✓ IDENTICAL (within 1e-6 tolerance)\n")
    } else if (max_diff < 0.01) {
      cat("  ✓ VERY CLOSE (within 0.01 tolerance)\n")
    } else {
      cat("  ⚠ DIFFERENCES DETECTED\n")

      # Show top differences
      top_idx <- order(diff, decreasing = TRUE)[1:min(5, length(diff))]
      cat("\n  Top differences:\n")
      for (idx in top_idx) {
        actual_idx <- which(both_valid)[idx]
        sp_name <- r_results$wfo_scientific_name[actual_idx]
        r_score <- r_val[actual_idx]
        py_score <- py_val[actual_idx]
        diff_val <- diff[idx]
        cat(sprintf("    %s: R=%.6f, Py=%.6f, diff=%.6f\n",
                    sp_name, r_score, py_score, diff_val))
      }
    }
  }
  cat("\n")
}

cat("============================================================\n")
cat("2. NaN SPECIES COMPARISON\n")
cat("============================================================\n\n")

r_nan <- r_results %>% filter(is.na(C))
py_nan <- py_results %>% filter(is.na(C))

cat(sprintf("R NaN species: %d\n", nrow(r_nan)))
cat(sprintf("Python NaN species: %d\n", nrow(py_nan)))

r_nan_set <- r_nan$wfo_taxon_id
py_nan_set <- py_nan$wfo_taxon_id

if (setequal(r_nan_set, py_nan_set)) {
  cat("✓ IDENTICAL: Same species fail in both implementations\n")
  cat("\nExamples (first 10):\n")
  for (name in head(r_nan$wfo_scientific_name, 10)) {
    cat(sprintf("  - %s\n", name))
  }
} else {
  cat("⚠ DIFFERENT: Different species fail\n")
  only_r <- setdiff(r_nan_set, py_nan_set)
  only_py <- setdiff(py_nan_set, r_nan_set)
  if (length(only_r) > 0) {
    cat(sprintf("\n  Only R fails (%d species)\n", length(only_r)))
  }
  if (length(only_py) > 0) {
    cat(sprintf("\n  Only Python fails (%d species)\n", length(only_py)))
  }
}

cat("\n")

cat("============================================================\n")
cat("3. ECOSYSTEM SERVICES COMPARISON\n")
cat("============================================================\n\n")

service_cols <- c(
  "npp_rating", "decomposition_rating", "nutrient_cycling_rating",
  "nutrient_retention_rating", "nutrient_loss_rating",
  "carbon_biomass_rating", "carbon_recalcitrant_rating",
  "carbon_total_rating", "erosion_protection_rating",
  "nitrogen_fixation_rating"
)

all_match <- TRUE
for (col in service_cols) {
  if (!(col %in% names(r_results)) || !(col %in% names(py_results))) {
    cat(sprintf("⚠ %s: Missing in one implementation\n", col))
    all_match <- FALSE
    next
  }

  r_val <- r_results[[col]]
  py_val <- py_results[[col]]

  # Handle NA matching
  match <- (r_val == py_val) | (is.na(r_val) & is.na(py_val))
  match_pct <- 100 * sum(match) / nrow(r_results)

  if (match_pct == 100) {
    cat(sprintf("✓ %s: 100%% match\n", col))
  } else {
    cat(sprintf("⚠ %s: %.2f%% match (%d differences)\n",
                col, match_pct, sum(!match)))
    all_match <- FALSE

    # Show examples
    diff_idx <- which(!match)[1:min(5, sum(!match))]
    cat("  Examples:\n")
    for (idx in diff_idx) {
      sp_name <- r_results$wfo_scientific_name[idx]
      r_val_ex <- r_val[idx]
      py_val_ex <- py_val[idx]
      cat(sprintf("    %s: R=%s, Py=%s\n", sp_name, r_val_ex, py_val_ex))
    }
  }
}

cat("\n")

cat("============================================================\n")
cat("FINAL VERDICT\n")
cat("============================================================\n\n")

# Overall check
csr_match <- TRUE
for (col in c("C", "S", "R")) {
  r_val <- r_results[[col]]
  py_val <- py_results[[col]]
  both_valid <- !is.na(r_val) & !is.na(py_val)
  if (sum(both_valid) > 0) {
    diff <- abs(r_val[both_valid] - py_val[both_valid])
    if (max(diff) > 0.01) {
      csr_match <- FALSE
    }
  }
}

nan_match <- setequal(r_nan_set, py_nan_set)

if (csr_match && nan_match && all_match) {
  cat("✓✓✓ PERFECT MATCH\n\n")
  cat("  • CSR scores identical (within floating-point precision)\n")
  cat("  • Same 30 species fail with NaN\n")
  cat("  • All ecosystem services match\n\n")
  cat("  R implementation is equivalent to Python implementation\n")
  cat("  Ready for Prof Shipley's review\n")
} else if (csr_match && nan_match) {
  cat("✓✓ CSR MATCH, ECOSYSTEM SERVICES DIFFER\n\n")
  cat("  • CSR scores identical\n")
  cat("  • Same edge cases\n")
  cat("  • Review ecosystem service logic for differences\n")
} else {
  cat("⚠ DISCREPANCIES FOUND\n\n")
  cat("  Review implementation differences above\n")
}

cat("\n")
